File: eseller_suite/eseller_suite/custom_script/stock_entry/stock_entry.js
frappe.ui.form.on('Stock Entry', {
	refresh(frm) {
		resrtict_actions_for_amazon_invoice(frm);
		set_filters(frm);
	}
})

frappe.ui.form.on('Stock Entry Detail', {
	item_code: function (frm, cdt, cdn) {
		fetch_item_details(frm, cdt, cdn);
	},
	qty: function (frm, cdt, cdn) {
		let row = locals[cdt][cdn];
		if (row.parentfield == "bundle_items") {
			handle_bundle_items(frm, cdt, cdn);
		}
	},
	basic_rate: function (frm, cdt, cdn) {
		let row = locals[cdt][cdn];
		if (row.parentfield == "bundle_items") {
			handle_bundle_items(frm, cdt, cdn);
		}
	},
	expense_account: function (frm, cdt, cdn) {
		let row = locals[cdt][cdn];
		if (row.parentfield == "bundle_items") {
			handle_bundle_items(frm, cdt, cdn);
		}
	},
	bundle_items_remove: function (frm, cdt, cdn) {
		handle_bundle_items(frm, cdt, cdn);
	},
	bundle_items_add: function (frm, cdt, cdn) {
		frappe.model.set_value(cdt, cdn, "s_warehouse", frm.doc.from_warehouse);
		frappe.model.set_value(cdt, cdn, "t_warehouse", frm.doc.to_warehouse);
	}
});

/**
 * Expands a bundle item into its component items in the Stock Entry.
 *
 * Fetches bundle components, removes previously linked rows,
 * recalculates quantities based on selected bundle qty,
 * and inserts updated component rows into the `items` table.
 */
function handle_bundle_items(frm, cdt, cdn) {
	frappe.call({
		method: "eseller_suite.eseller_suite.custom_script.stock_entry.stock_entry.process_bundle_items",
		args: {
			bundle_items: JSON.stringify(frm.doc.bundle_items)
		},
		callback: function (r) {
			if (!r.message) return;
			frm.clear_table("items");
			frm.refresh_field("items");
			r.message.forEach(function (bundle_child) {
				frm.add_child("items", bundle_child);
			});
			frm.refresh_field("items");
		}
	});
	frm.refresh_field("items");
}

function resrtict_actions_for_amazon_invoice(frm) {
	if (frm.doc.amazon_invoice_id) {
		frm.disable_form();
		frm.disable_save();
		frm.clear_custom_buttons();
	}
}

/*
* Function to apply filters in the item_code field in Bundle Items child table
 */
function set_filters(frm) {
	frm.set_query("item_code", "bundle_items", function () {
		return {
			filters: {
				is_bundle_item: 1
			}
		};
	});
	frm.set_query("expense_account", "bundle_items", function () {
		return {
			filters: {
				company: frm.doc.company,
			}
		};
	});
	frm.set_query("s_warehouse", "bundle_items", function () {
		return {
			filters: {
				company: frm.doc.company,
			}
		};
	});
	frm.set_query("t_warehouse", "bundle_items", function () {
		return {
			filters: {
				company: frm.doc.company,
			}
		};
	});

}

function fetch_item_details(frm, cdt, cdn) {
	let row = locals[cdt][cdn];
	if (row.item_code) {
		frappe.call({
			method: "frappe.client.get",
			args: {
				doctype: "Item",
				name: row.item_code
			},
			callback: function (r) {
				if (!r.message) return;
				let item = r.message;
				frappe.model.set_value(cdt, cdn, "item_name", item.item_name);
				frappe.model.set_value(cdt, cdn, "description", item.description);
				frappe.model.set_value(cdt, cdn, "uom", item.stock_uom);
			}
		});
	}
}

File: eseller_suite/eseller_suite/custom_script/stock_entry/stock_entry.py
import frappe
from frappe.model import _
from frappe.model.base_document import flt

def validate(doc, method):
	'''
		Method which trigger on validate event of Stock Entry
	'''
	set_bundle_diff_amount(doc)

def on_submit(doc, method):
	'''
		Method which trigger on on_submit event of Stock Entry
	'''
	validate_bundle_amount_difference(doc)

def transfer_barcodes(doc, method=None):
	"""method transfers the barcodes on submit

	Args:
		doc (_type_): _description_
		method (_type_, optional): _description_. Defaults to None.
	"""
	for item in doc.items:
		if not item.barcode_no:
			continue

		barcodes = item.barcode_no.split("\n")
		for barcode in barcodes:
			existing_serial_no = frappe.db.exists(
				"eSeller Serial No",
				{"serial_no": barcode},
			)
			if existing_serial_no:
				serial_doc = frappe.get_doc("eSeller Serial No", existing_serial_no)
				serial_doc.status = "Transferred"
				serial_doc.warehouse = item.t_warehouse
				serial_doc.transfer_document_no = doc.name
				serial_doc.save()
	frappe.msgprint(
		"Serial Numbers are now Transferred",
		alert=True,
	)

def before_insert_custom(doc, method=None):
	"""method corrects warehouse in stock returns from returns & replaced orders if temporary stock transfer is on"""
	if doc.from_return_invoice:
		amz_settings = frappe.get_last_doc("Amazon SP API Settings")
		if amz_settings.temporary_stock_transfer_required:
			si = doc.sales_invoice_no
			return_warehouse = amz_settings.warehouse
			si_fulfillment_channel = frappe.db.get_value("Sales Invoice", si, "fulfillment_channel")

			if si_fulfillment_channel:
				if si_fulfillment_channel == 'AFN':
					return_warehouse = amz_settings.afn_warehouse

			doc.to_warehouse = return_warehouse
			for row in doc.items:
				row.t_warehouse = return_warehouse

def on_canel(doc, method):
	doc.ignore_linked_doctypes = (
		"GL Entry",
		"Stock Ledger Entry",
		"Repost Item Valuation",
		"Serial and Batch Bundle",
		"Amazon STN Entry",
	)

def set_bundle_diff_amount(doc):
	'''
		Method to set bundle differences and total
	'''
	#Calculating based on Items table
	total_bundle_amount = 0
	for row in doc.items:
		if row.from_bundle_item:
			total_bundle_amount += row.amount
	doc.total_bundle_amount = round(total_bundle_amount, 2)

	#Calculating based on Bundle Items Table
	total_bundle_amount_actual = 0
	for item in doc.bundle_items:
		if item.basic_rate and item.qty:
			item.amount = item.basic_rate * item.qty
		total_bundle_amount_actual += item.amount
	doc.total_bundle_amount_actual = round(total_bundle_amount_actual, 2)

	# Setting Difference Amount
	doc.bundle_difference_amount = round((doc.total_bundle_amount_actual - doc.total_bundle_amount), 2)

def validate_bundle_amount_difference(doc):
	"""
	Method to validate bundle difference amount on submission of stock entry
	"""
	if doc.bundle_difference_amount:
		title = 'Check Difference Amount'
		msg = 'Cannot submit the stock entry due to difference amount of {0} for bundle items'.format(frappe.bold(doc.bundle_difference_amount))
		frappe.throw(title=title, msg=msg)

@frappe.whitelist()
def get_bundle_items(bundle_item):
	"""
	method fetches the child items of a bundle item to be added in the stock entry items table
	"""
	if not bundle_item:
		return []

	if not frappe.db.exists("Product Bundle", bundle_item):
		frappe.log_error(_("Product Bundle {0} does not exist").format(bundle_item))
		return []

	bundle_items = frappe.get_all(
		"Product Bundle Item",
		filters={"parent": bundle_item},
		fields=["item_code", "qty", "description"],
		order_by="idx"
	)
	result = []
	for row in bundle_items:
		item_details = frappe.get_cached_value(
			"Item",
			row.item_code,
			["item_name", "stock_uom", "standard_rate"],
			as_dict=True
		)
		result.append({
			"item_code": row.item_code,
			"item_name": item_details.item_name if item_details else "",
			"qty": flt(row.qty),
			"uom": item_details.stock_uom,
			"rate": item_details.standard_rate if item_details else 0,
			"description": row.description
		})
	return result

